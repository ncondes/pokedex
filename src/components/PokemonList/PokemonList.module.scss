@use '@/scss/mixins' as *;

.pokemon_grid {
   width: 100%;
   display: grid;
   place-items: center;
   grid-template-columns: repeat(1, minmax(280px, 1fr));
   gap: 2.5rem;

   @include breakpoint-up('md') {
      grid-template-columns: repeat(2, minmax(280px, 1fr));
   }

   @include breakpoint-up('lg') {
      grid-template-columns: repeat(3, minmax(280px, 1fr));
   }

   @include breakpoint-up('xl') {
      grid-template-columns: repeat(4, minmax(280px, 1fr));
   }
}

.no_results {
   font-size: 2rem;
   font-weight: 600;
   text-align: center;
}
