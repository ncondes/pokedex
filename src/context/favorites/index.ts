export { FavoritesContext } from './FavoritesContext'
export { FavoritesProvider } from './FavoritesProvider'
