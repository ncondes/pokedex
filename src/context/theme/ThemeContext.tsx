'use client'

import { createContext } from 'react'

interface Props {
   theme: 'light' | 'dark'
   toggleTheme: () => void
   useTheme: (className: string) => string
}

export const ThemeContext = createContext({} as Props)
