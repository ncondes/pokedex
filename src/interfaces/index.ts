export * from './data'
export * from './pokemon'
