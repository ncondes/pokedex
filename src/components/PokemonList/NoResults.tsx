'use client'
import { useContext } from 'react'
import styles from './PokemonList.module.scss'
import { ThemeContext } from '@/context/theme'

export const NoResults = () => {
   const { useTheme } = useContext(ThemeContext)

   return (
      <p className={`${styles.no_results} ${useTheme('font')}`}>
         No Pokemons Found.
      </p>
   )
}
