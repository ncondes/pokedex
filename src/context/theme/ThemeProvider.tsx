'use client'

import { FC, PropsWithChildren, useEffect, useState } from 'react'
import { ThemeContext } from './ThemeContext'
import { getLocalTheme, setLocalTheme } from '@/utils'

export const ThemeProvider: FC<PropsWithChildren> = ({ children }) => {
   const [theme, setTheme] = useState<'light' | 'dark'>('light')

   const toggleTheme = () => {
      const t = theme === 'light' ? 'dark' : 'light'
      setTheme(t)
      setLocalTheme(t)
   }

   const useTheme = (className: string): string => `${theme}-${className}`

   useEffect(() => {
      const initialTheme = getLocalTheme() as 'light' | 'dark'
      setTheme(initialTheme)
   }, [])

   return (
      <ThemeContext.Provider value={{ theme, toggleTheme, useTheme }}>
         {children}
      </ThemeContext.Provider>
   )
}
