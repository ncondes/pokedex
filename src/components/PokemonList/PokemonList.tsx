import { FC } from 'react'
import { Pokemon } from '@/interfaces'
import { NoResults } from './NoResults'
import styles from './PokemonList.module.scss'

interface Props {
   pokemons: Pokemon[]
}

export const PokemonList: FC<Props> = ({ pokemons }) => {
   return (
      <>
         {pokemons.length === 0 ? (
            <NoResults />
         ) : (
            <div className={styles.pokemon_grid}>
               {pokemons.map((pokemon) => (
                  <p key={pokemon.id}>{pokemon.name}</p>
               ))}
            </div>
         )}
      </>
   )
}
