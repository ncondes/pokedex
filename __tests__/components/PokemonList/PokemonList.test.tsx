import { PokemonList } from '@/components/PokemonList'
import { ThemeProvider } from '@/context/theme'
import { Pokemon, Stat } from '@/interfaces'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'

describe('PokemonList', () => {
   it('renders the correct number of PokemonItem components', () => {
      const pokemons = [
         {
            id: 1,
            name: 'Pikachu',
            types: [{ type: { name: 'electric' } }],
            sprites: {
               front_default:
                  'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/25.png',
            },
            stats: [] as Stat[],
         },
         {
            id: 2,
            name: 'Charizard',
            types: [{ type: { name: 'fire' } }],
            sprites: {
               front_default:
                  'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/6.png',
            },
            stats: [] as Stat[],
         },
      ] as Pokemon[]

      const { getByText } = render(
         <ThemeProvider>
            <PokemonList pokemons={pokemons} />
         </ThemeProvider>
      )

      const pikachu = getByText('Pikachu')
      const charizard = getByText('Charizard')

      expect(pikachu).toBeInTheDocument()
      expect(charizard).toBeInTheDocument()
   })

   it('renders the NoResults component when there are no pokemons', () => {
      const pokemons = [] as Pokemon[]

      const { getByText } = render(
         <ThemeProvider>
            <PokemonList pokemons={pokemons} />
         </ThemeProvider>
      )

      const noResults = getByText('No Pokemons Found.')

      expect(noResults).toBeInTheDocument()
   })

   it('should match the snapshot', () => {
      const pokemons = [
         {
            id: 1,
            name: 'Pikachu',
            types: [{ type: { name: 'electric' } }],
            sprites: {
               front_default:
                  'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/25.png',
            },
            stats: [] as Stat[],
         },
         {
            id: 2,
            name: 'Charizard',
            types: [{ type: { name: 'fire' } }],
            sprites: {
               front_default:
                  'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/6.png',
            },
            stats: [] as Stat[],
         },
      ] as Pokemon[]

      const { container } = render(
         <ThemeProvider>
            <PokemonList pokemons={pokemons} />
         </ThemeProvider>
      )

      expect(container.firstChild).toMatchSnapshot()
   })
})
