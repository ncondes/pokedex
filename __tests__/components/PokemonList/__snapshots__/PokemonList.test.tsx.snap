// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonList should match the snapshot 1`] = `
<div
  class="pokemon_grid"
>
  <article
    class="pokemon_card pokemon-type-electric"
  >
    <section>
      <div
        class="pokemon_header"
      >
        <h2>
          Pikachu
        </h2>
        <button>
          <i
            class="fa-regular fa-heart fa-xl"
          />
        </button>
      </div>
      <div
        class="pokemon_chips"
      >
        <div
          class="pokemon_chip"
        >
          Electric
        </div>
      </div>
      <div
        class="pokemon_bg_image_container"
      >
        <div
          class="pokeball_icon"
        >
          <img
            alt="pokeball icon"
            data-nimg="1"
            decoding="async"
            height="150"
            loading="lazy"
            src="/_next/image?url=%2Fimages%2Fpokeball-icon.png&w=384&q=75"
            srcset="/_next/image?url=%2Fimages%2Fpokeball-icon.png&w=256&q=75 1x, /_next/image?url=%2Fimages%2Fpokeball-icon.png&w=384&q=75 2x"
            style="color: transparent;"
            width="150"
          />
        </div>
        <div
          class="points_icon"
        >
          <img
            alt="points icon"
            data-nimg="1"
            decoding="async"
            height="50"
            loading="lazy"
            src="/_next/image?url=%2Fimages%2Fpoints-icon.png&w=128&q=75"
            srcset="/_next/image?url=%2Fimages%2Fpoints-icon.png&w=64&q=75 1x, /_next/image?url=%2Fimages%2Fpoints-icon.png&w=128&q=75 2x"
            style="color: transparent;"
            width="50"
          />
        </div>
        <div
          class="lines_icon"
        >
          <img
            alt="lines icon"
            data-nimg="1"
            decoding="async"
            height="50"
            loading="lazy"
            src="/_next/image?url=%2Fimages%2Flines-icon.png&w=128&q=75"
            srcset="/_next/image?url=%2Fimages%2Flines-icon.png&w=64&q=75 1x, /_next/image?url=%2Fimages%2Flines-icon.png&w=128&q=75 2x"
            style="color: transparent;"
            width="50"
          />
        </div>
        <div
          class="pokemon_image"
        >
          <img
            alt="pokemon image"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="200"
            src="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F25.png&w=640&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F25.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F25.png&w=640&q=75 2x"
            style="color: transparent;"
            width="200"
          />
        </div>
      </div>
    </section>
    <div
      class="pokemon_stats light-card-bg"
    >
      <p
        class="light-font"
      >
        Base Stats
      </p>
      <div
        class="base_stats"
      />
    </div>
  </article>
  <article
    class="pokemon_card pokemon-type-fire"
  >
    <section>
      <div
        class="pokemon_header"
      >
        <h2>
          Charizard
        </h2>
        <button>
          <i
            class="fa-regular fa-heart fa-xl"
          />
        </button>
      </div>
      <div
        class="pokemon_chips"
      >
        <div
          class="pokemon_chip"
        >
          Fire
        </div>
      </div>
      <div
        class="pokemon_bg_image_container"
      >
        <div
          class="pokeball_icon"
        >
          <img
            alt="pokeball icon"
            data-nimg="1"
            decoding="async"
            height="150"
            loading="lazy"
            src="/_next/image?url=%2Fimages%2Fpokeball-icon.png&w=384&q=75"
            srcset="/_next/image?url=%2Fimages%2Fpokeball-icon.png&w=256&q=75 1x, /_next/image?url=%2Fimages%2Fpokeball-icon.png&w=384&q=75 2x"
            style="color: transparent;"
            width="150"
          />
        </div>
        <div
          class="points_icon"
        >
          <img
            alt="points icon"
            data-nimg="1"
            decoding="async"
            height="50"
            loading="lazy"
            src="/_next/image?url=%2Fimages%2Fpoints-icon.png&w=128&q=75"
            srcset="/_next/image?url=%2Fimages%2Fpoints-icon.png&w=64&q=75 1x, /_next/image?url=%2Fimages%2Fpoints-icon.png&w=128&q=75 2x"
            style="color: transparent;"
            width="50"
          />
        </div>
        <div
          class="lines_icon"
        >
          <img
            alt="lines icon"
            data-nimg="1"
            decoding="async"
            height="50"
            loading="lazy"
            src="/_next/image?url=%2Fimages%2Flines-icon.png&w=128&q=75"
            srcset="/_next/image?url=%2Fimages%2Flines-icon.png&w=64&q=75 1x, /_next/image?url=%2Fimages%2Flines-icon.png&w=128&q=75 2x"
            style="color: transparent;"
            width="50"
          />
        </div>
        <div
          class="pokemon_image"
        >
          <img
            alt="pokemon image"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="200"
            src="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F6.png&w=640&q=75"
            srcset="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F6.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F6.png&w=640&q=75 2x"
            style="color: transparent;"
            width="200"
          />
        </div>
      </div>
    </section>
    <div
      class="pokemon_stats light-card-bg"
    >
      <p
        class="light-font"
      >
        Base Stats
      </p>
      <div
        class="base_stats"
      />
    </div>
  </article>
</div>
`;
